jersey-servers/src/main/java/com/palantir/remoting3/servers/jersey/QosExceptionMapper.java

```diff
@@ -0,0 +1,89 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.remoting3.servers.jersey;
+
+import com.google.common.net.HttpHeaders;
+import com.palantir.remoting.api.errors.QosException;
+import java.net.URL;
+import java.time.Duration;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An {@link ExceptionMapper} that turns http-remoting {@link QosException}s into appropriate HTTP error codes and
+ * headers. Four different cases are distinguished:
+ * <ol>
+ * <li>Retry any node of this service immediately: HTTP 307 Temporary Redirect</li>
+ * <li>Retry any node of this service after a given backoff time: HTTP 429 Too Many Requests + Retry-After header</li>
+ * <li>Retry a specific (other) node of this service immediately: HTTP 308 Permanent Redirect + Location header</li>
+ * <li>Don't retry any node of this service: HTTP 503 Unavailable</li>
+ * </ol>
+ */
+@Provider
+final class QosExceptionMapper implements ExceptionMapper<QosException> {
+
+    private static final Logger log = LoggerFactory.getLogger(QosExceptionMapper.class);
+
+    @Override
+    public Response toResponse(QosException exception) {
+        log.debug("Possible quality-of-service intervention", exception);
+        if (exception instanceof QosException.Retry) {
+            QosException.Retry retry = (QosException.Retry) exception;
+            if (!retry.getBackoff().isPresent() && !retry.getRedirectTo().isPresent()) {
+                return retryAnyNow();
+            } else {
+                if (retry.getBackoff().isPresent()) {
+                    return retryAnyLater(retry.getBackoff().get());
+                } else if (retry.getRedirectTo().isPresent()) {
+                    return retryThisHostNow(retry.getRedirectTo().get());
+                } else {
+                    throw new RuntimeException("Unexpected internal state: expected either backoff or redirectTo");
+                }
+            }
+        } else if (exception instanceof QosException.Unavailable) {
+            return unavailable();
+        } else {
+            throw new RuntimeException(
+                    "Internal error, unknown QosException subclass: " + exception.getClass().getSimpleName());
+        }
+    }
+
+    private Response retryAnyNow() {
+        return Response.status(307).build();
+    }
+
+    private Response retryAnyLater(Duration backoff) {
```

Comment:
yeah so, I wanna be able to send a 429 without a RETRY_AFTER header 

this is valid http...

> The response representations SHOULD include details explaining the condition, and MAY include a Retry-After header indicating how long to wait before making a new request.